### DISCUSSION TITLE
upload csv File from node

### DISCUSSION THREAD

**Discussion by lupeaserban** [urql-graphql/urql#3179] [Q&A]
I have graphql endpoint that accepts the scalar type Upload and is expecting a csv File. 

I've tried ( different variations of ) the following code:

```
  const mutation = gql`
    mutation importStuff($csvFile: Upload!) {
      importStuff(csvFile: $csvFile) {
        errors
        processed
        imported
      }
    }
  `;

  const fileBuffer = fs.readFileSync(pathToFile);

  const file = new Blob([fileBuffer], {
    type: 'text/csv',
  });

  const vars: any = {
    csvFile: file,
  };

  const resp = await client.mutation(mutation, vars);
```

..to which i get a GraphQl error with a nested message: BAD_USER_INPUT

I'm pretty sure the error is because the `file` I'm sending is not of type 'Upload' but have no ideea how to proceed. 

Any help or lead is appreciated :D

**Comment by kitten** [comment]
It's hard for me to really pinpoint the issue here, but you'd have to validate that:
- the multipart request is sent correctly
- the multipart request is received and parsed correctly
- and; that you're on the latest version of `@urql/core`

Depending on the fetch polyfill, maybe the request isn't working as you'd expect. Alternatively, it's possible that the API isn't quite set up to parse the body and file stream correctly 

Also, as a hint, scalars are a bit of a funny thing and they don't work as exact types. It's really up to the API to parse and accept them correctly. However, multipart uploads are a bit of a special thing since they are a spec extension. So, there isn't an `Upload` scalar of a certain type per se, but the API sees the multipart request and fills in the type as an incoming multipart stream

**Reply by lupeaserban** [reply]
Thanks for looking into this.

I'm on the latest version of `@urql/core` and I'm using `cross-fetch` as a polyfill. The API is setup correctly because im able to use the endpoint from a client like Altair/Postman. 

This is how I'm instantiating the client in my app:

```
 const client = new Client({
    url: 'http://localhost:3000/admin-api',
    exchanges: [fetchExchange],
    fetch,
  });
```

Considering this phrase from the docs: 
_If a GraphQL API supports this, we can pass a [File](https://developer.mozilla.org/en-US/docs/Web/API/File) or a [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob) directly into our variables and define the corresponding scalar for our variable, which is often called File or Upload._

Would you do anything different than I am?
